### PR TITLE
fix(aws-kinesis): remove default shard count when stream mode is on-demand and set default mode to provisioned

### LINKED ISSUE #18139: (aws-kinesis): fails to deploy on-demand stream
### What is the problem?

Error from CloudFormation when using Kinesis Data Streams in on-demand mode with shardCount specified.

> ShardCount is not expected when StreamMode=ON_DEMAND

### Reproduction Steps

Create a new stream with a `streamMode` set.

```typescript
new Stream(this, 'KinesisStream', {
    streamName: 'my-ingest-stream',
    shardCount: 1, // optional
    streamMode: StreamMode.ON_DEMAND,
});
```

### What did you expect to happen?

Deploy a stream in on-demand mode, possibly informed by the shardCount for an initial scaling hint.

### What actually happened?

Error from CloudFormation

> ShardCount is not expected when StreamMode=ON_DEMAND

### CDK CLI Version

1.137.0 (build bfbdf64)

### Framework Version

1.137.0

### Node.js Version

14.17.3

### OS

MacOS

### Language

Typescript

### Language Version

4.5.2

### Other information

[CloudFormation's documentation][docs] does not inform the user that setting the StreamMode to ON_DEMNAD is an error. Can we raise a bug with the right team to get the documentation improved?

Removing the `shardCount` field from the CDK user code does not fix the error because internally, CDK always sets the shardCount to 1 if it is left undefined by the user.

The solution in CDK is to remove `shardCount` from the underlying `CfnStream` if the user did not provide the field. CloudFormation will default to 1 anyway if needed, so CDK's default is not doing anything useful here.

We should also add a validator to ensure that `shardCount` and `StreamMode: 'ON_DEMAND'` are not set at the same time when generating the template.

An end-to-end test through CloudFormation would have caught this bug before release. I'm not aware that there is any automated testing of this type at present.

[docs]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-kinesis-stream.html

### PR DESCRIPTION
**Pull request by igilham**
Change the default Kinesis Data Stream's stream mode to provisioned from undefined to make the active configuration more explicit in the resulting CloudFormation templates.

Fix an issue whereby the shard count is always set when the stream mode is set to on-demand, which is invalid. Shard count still defaults to `1` in provisioned mode, but is left undefined in on-demand mode.

Add validation for the above so that an error is thrown from CDK when specifying on-demand mode with a shard count.

Fixes #18139

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by igilham** [comment]
The changes so far should fix the basic problem and allow the stream mode to be used.

I haven't figured out how to do validation yet so any help would be appreciated.

**Comment by igilham** [comment]
It looks like the dynamodb module just adds various guard statements and to the L2 construct constructor to prevent misuse of conflicting configuration options. I'll do the same.

**Comment by igilham** [comment]
Okay, I think I'm ready for the first round of reviews.

File: packages/@aws-cdk/aws-kinesis/test/stream.test.ts [hunk 1]
```diff
@@ -431,6 +511,17 @@ describe('Kinesis data streams', () => {
...
 
+  test('throws when using shardCount with on-demand streamMode', () => {
+    const stack = new Stack();
+
+    expect(() => {
+      new Stream(stack, 'MyStream', {
+        shardCount: 2,
+        streamMode: StreamMode.ON_DEMAND,
+      });
+    }).toThrow('streamMode must be set to PROVISIONED (default) when specifying shardCount');
```
Comment (d4r3topk) [inline packages/@aws-cdk/aws-kinesis/test/stream.test.ts:522]: You can use `streamMode must be set to ${StreamMode.PROVISIONED} (default) when specifying shardCount` to keep the value of the enum even if it is changed in the future.

File: packages/@aws-cdk/aws-kinesis/test/integ.stream-dashboard.expected.json [hunk 2]
```diff
@@ -4,6 +4,9 @@
       "Type": "AWS::Kinesis::Stream",
       "Properties": {
         "ShardCount": 1,
+        "StreamModeDetails": {
+          "StreamMode": "PROVISIONED"
```
Comment (d4r3topk) [inline packages/@aws-cdk/aws-kinesis/test/integ.stream-dashboard.expected.json:8]: Check for ${StreamMode.PROVISIONED} for the exact enum values and same for ON_DEMAND occurences

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts [hunk 3]
```diff
@@ -752,9 +755,15 @@ export class Stream extends StreamBase {
...
     });
 
-    const shardCount = props.shardCount || 1;
+    let shardCount = props.shardCount;
+    let streamMode = props.streamMode ?? StreamMode.PROVISIONED;
 
-    const streamMode = props.streamMode;
+    if (streamMode === StreamMode.ON_DEMAND && shardCount !== undefined) {
+      throw new Error('streamMode must be set to PROVISIONED (default) when specifying shardCount');
+    }
```
Comment (d4r3topk) [inline packages/@aws-cdk/aws-kinesis/lib/stream.ts:763]: Could be an else if

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts [hunk 4]
```diff
@@ -752,9 +755,15 @@ export class Stream extends StreamBase {
       physicalName: props.streamName,
     });
 
-    const shardCount = props.shardCount || 1;
+    let shardCount = props.shardCount;
+    let streamMode = props.streamMode ?? StreamMode.PROVISIONED;
```
Comment (d4r3topk) [inline packages/@aws-cdk/aws-kinesis/lib/stream.ts:759]: Can be constant

Comment (d4r3topk) [reply packages/@aws-cdk/aws-kinesis/lib/stream.ts:759]: In this commit, you are making two changes, ie. the fix and the default to PROVISIONED capacity. You can include that in the commit details.

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts (hunk 3 above)
Comment (igilham) [reply packages/@aws-cdk/aws-kinesis/lib/stream.ts:763]: This is a matter of personal preference. Adding `else` here doesn't make this code any clearer to me so I prefer not to.

File: packages/@aws-cdk/aws-kinesis/test/integ.stream-dashboard.expected.json (hunk 2 above)
Comment (igilham) [reply packages/@aws-cdk/aws-kinesis/test/integ.stream-dashboard.expected.json:8]: I've made this change in the unit tests. It's not applicable to these JSON files.

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts (hunk 4 above)
Comment (igilham) [reply packages/@aws-cdk/aws-kinesis/lib/stream.ts]: I've updated the description. Hopefully that's a bit clearer.

**Comment by igilham** [comment]
I'm back from today if there's any more feedback on this.

**Review by peterwoodworth — APPROVED** [review]
This looks great to me, I wish I saw this before I submitted my own PR! Thank you for the contribution 🙂 

Thank you also to @d4r3topk for the initial review!

**Pull request merged**